### DISCUSSION TITLE
Jupyter widget

### DISCUSSION THREAD

**Discussion by hadim** [molstar/molstar#144] [Ideas]
Is there is any plan to make a Jupyter plugin out of this library?

Similar to https://github.com/arose/nglview

**Comment by arose** [comment]
Hi @hadim, we don't have immediate plans to do it ourselves. It would be great to see it though.

**Reply by krassowski** [reply]
If anyone would like to embark on the journey I am happy to assist with JupyterLab integration (but I don't have bandwidth to write anything myself this year).

CC @hainm

**Reply by hainm** [reply]
I don't have bandwidth either. It seems it's a non-trivial work.

**Comment by dsehnal** [comment]
I think I can put up an MVP for this with minimal functionality and maybe someone will help pick this up from there.

**Reply by hainm** [reply]
@dsehnal  That will help a lot. And it's actually the most difficult task for me since I have no idea how to use `molstar`'s API (the library is huge).

**Comment by hainm** [comment]
@dsehnal hi David, I start playing with the widget for molstar but getting error. This simple line of code: https://github.com/hainm/molstarview/blob/58c98ecbdd49fdad5c84ce3c3000740d055e7437/js/lib/widget.js#L3

cause `Uncaught (in promise) Error: Automatic publicPath is not supported in this browser` in jupyter notebook. 
(Using `import` does not help either :D). 

Here is the content of webpack.config.js: https://github.com/hainm/molstarview/blob/master/js/webpack.config.js

Do you have any idea why? 

Those are steps to reproduce:
```
git clone https://github.com/hainm/molstarview
cd molstarview 
pip install . 
jupyter-notebook
```
In the notebook, use those commands:
```
from molstarview.widget import MolstarView

v = MolstarView()
v
```

**Reply by hainm** [reply]
By the way, we did the same thing for `NGL`: https://github.com/nglviewer/nglview/blob/fb3531b2999f603841488b680cb8665ae63e6870/js/src/widget_ngl.ts#L3

**Reply by dsehnal** [reply]
Hi @hainm, you are importing the Viewer which is already built for the browser. ``require(molstar/lib/apps/viewer/app)`` is what you want for your test case. This will have the ``Viewer`` class. Altho I don't think the widget should be based on this instance.

**Reply by hainm** [reply]
> Altho I don't think the widget should be based on this instance.

@dsehnal Can you please provide a minimal example?

**Reply by hainm** [reply]
By the way, I got another failure if following your suggestion:

```
ERROR in ./node_modules/h264-mp4-encoder/embuild/dist/h264-mp4-encoder.node.js 1:1020261-1020278
Module not found: Error: Can't resolve 'crypto' in '/Users/nguyen/programs/molstarview/js/node_modules/h264-mp4-encoder/embuild/dist'

BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
This is no longer the case. Verify if you need this module and configure a polyfill for it.

If you want to include a polyfill, you need to:
	- add a fallback 'resolve.fallback: { "crypto": require.resolve("crypto-browserify") }'
	- install 'crypto-browserify'
If you don't want to include a polyfill, you can use an empty module like this:
	resolve.fallback: { "crypto": false }
 @ ./node_modules/molstar/lib/extensions/mp4-export/encoder.js 7:0-40 23:41-65
 @ ./node_modules/molstar/lib/extensions/mp4-export/controls.js 11:0-47 65:45-63
 @ ./node_modules/molstar/lib/extensions/mp4-export/ui.js 15:0-61 48:59-70 73:869-887
 @ ./node_modules/molstar/lib/extensions/mp4-export/index.js 8:0-36 21:63-75
 @ ./node_modules/molstar/lib/apps/viewer/app.js 17:0-56 51:38-47
 @ ./lib/widget.js
 @ ./lib/index.js 2:0-39
```

PS: I have limited knowledge about Javascript.

**Reply by hainm** [reply]
By the way, the [docs[(https://molstar.org/docs/plugin/) says:
```
When using WebPack (or possibly other build tool) with the Mol* NPM package installed, the viewer class can be imported using


import { Viewer } from 'molstar/build/viewer/molstar'

function initViewer(target: string | HTMLElement) {
    return new Viewer(target, { /* options */})
}
```

The jupyter widget is using webpack.

**Reply by dsehnal** [reply]
Yes, this is one way of doing that. I just don't think the `Viewer` wrapper is suited for integration with Jupyter Widger (tho ofc I might be wrong and you could be able to use it).

Something like this could work (sorry I didn't test the code). Might require modification of the webpack file to support ``async``, tho I guess this could be rewritten with ``.then`` too.

```ts
import * as widgets from '@jupyter-widgets/base';
import { createPluginUI } from 'molstar/lib/mol-plugin-ui';
import { DownloadStructure, PdbDownloadProvider } from 'molstar/lib/mol-plugin-state/actions/structure';

const MolstarModel = widgets.DOMWidgetModel.extend({
    defaults: _.extend(widgets.DOMWidgetModel.prototype.defaults(), {
        _model_name : 'MolstarModel',
        _view_name : 'MolstarView',
        _model_module : 'jupyter-widget-example',
        _view_module : 'jupyter-widget-example',
        _model_module_version : '0.1.0',
        _view_module_version : '0.1.0',
        value : 'Molstar World!'
    })
});

const MolstarView = widgets.DOMWidgetView.extend({
    // Defines how the widget gets rendered into the DOM
    render: async function() {       
        this.plugin = await createPluginUI(this.el);
        
        // call it after the plugin has been initialized
        this.value_changed();
        this.model.on('change:value', this.value_changed, this);
    },

    value_changed: function() {
        this.loadPdb(this.model.get('value'));
    },
    
    loadPdb(pdb: string) {    
        // this method is taken from the Viewer class
        const params = DownloadStructure.createDefaultParams(this.plugin.state.data.root.obj, this.plugin);
        const provider = this.plugin.config.get(PluginConfig.Download.DefaultPdbProvider);
        return this.plugin.runTask(this.plugin.state.data.applyAction(DownloadStructure, {
            source: {
                name: 'pdb' as const,
                params: {
                    provider: {
                        id: pdb,
                        server: {
                            name: provider,
                            params: PdbDownloadProvider[provider].defaultValue as any
                        }
                    },
                    options: { ...params.source.params.options },
                }
            }
        }));
    }
});
```

**Reply by hainm** [reply]
Thanks @dsehnal,

I am able to import `molstar` to Jupyter notebook after editing your code.
The view, however, is frozen. Means the buttons are not clickable and nothing really works.

But we made progress on this. :D 

<img width="1132" alt="image" src="https://user-images.githubusercontent.com/4451957/172401574-d24bfff3-10e4-48b4-8448-15ad4bf3aab9.png">

**Reply by dsehnal** [reply]
Right, this also needs CSS: ``import molstar/lib/mol-plugin-ui/skin/light.scss`` + needs SCSS loader in WebPack. Alternatively, importing ``'molstar/build/viewer/molstar.css'`` will also work (it's the built version of ``light.scss``).

When I did a proprietary implementation of Mol* in Jupyter, I used Material-UI for custom UI with CSS in JS approach, so that is also an option, but more work.

**Reply by dsehnal** [reply]
Also;

- You can hide the UI elements by default, by specifying a second parameter to ``createPluginUI`` (look up ``PluginUISpec``).
- Maybe ``this.el`` needs some styling, like giving it specific width/height. Might actually have to do something like:

```ts
const container = document.createElement('div');
container.style.width = '800px';
container.style.height = '600px';
this.el.appendChild(container);
await createPluginUI(container);
```

**Reply by hainm** [reply]
Great. Thanks David again. Will try and ping you later. Cheers.

**Reply by hainm** [reply]
> Right, this also needs CSS: import molstar/lib/mol-plugin-ui/skin/light.scss + needs SCSS loader in WebPack.

That helps

- Running the first cell will result this
<img width="1259" alt="image" src="https://user-images.githubusercontent.com/4451957/172520647-ce58f04b-d8e7-4620-bb1a-3bb1578bcc66.png">

- Click the setting icon will result this
<img width="1270" alt="image" src="https://user-images.githubusercontent.com/4451957/172520686-fee06350-cd83-4bba-8b9a-6313056d53b9.png">

- Executing the 2nd notebook cell (`v.value = '1l2y'`) will result this (the number of pdb models are correct, `38` for `1l2y`)
<img width="1271" alt="image" src="https://user-images.githubusercontent.com/4451957/172520750-8ab36403-823a-4302-bc71-081d803897db.png">

**Reply by hainm** [reply]
- After clicking the fullscreen (?) button and clicking the style:

<img width="1344" alt="image" src="https://user-images.githubusercontent.com/4451957/172523346-7a2c1be0-cc7b-4690-a489-41b6c61e9b84.png">

**Reply by dsehnal** [reply]
Try calling ``this.plugin.handleResize()`` in ``value_changed``. Perhaps add ``await new Promise(res => requestAnimationFrame(() => plugin.handleResize()))`` after the create too.

In my implementation I used React to mount to ``this.el`` and didn't have this problem due to the lifecycle hooks there (e.g. the plugin was created only after all the DOM elements were property mounted etc.)

**Reply by hainm** [reply]
> (e.g. the plugin was created only after all the DOM elements were property mounted etc.)

Thanks @dsehnal. I've moved the block of codes to `this.displayed(...)` and that helps: https://github.com/hainm/molstarview/blob/187452c804bf8314b0e81a74bf3dccfeedcb22c9/js/lib/widget.js#L44-L57

**Reply by hainm** [reply]
hi @dsehnal 
My next question: What's the easiest way to update coordinates for a pdb structure?
In `NGL`, I could do something like this:
https://github.com/nglviewer/nglview/blob/fb3531b2999f603841488b680cb8665ae63e6870/js/src/widget_ngl.ts#L886-L896

PS: I've read https://github.com/molstar/molstar/issues/282 but it seems not to be trivial to me. 😅
Thanks.

**Reply by hainm** [reply]
> hi @dsehnal My next question: What's the easiest way to update coordinates for a pdb structure? In `NGL`, I could do something like this: https://github.com/nglviewer/nglview/blob/fb3531b2999f603841488b680cb8665ae63e6870/js/src/widget_ngl.ts#L886-L896
> 
> PS: I've read #282 but it seems not to be trivial to me. 😅 Thanks.

Something likes below? (from: https://github.com/molstar/molstar/blob/d1e17785b8404eec280ad04a6285ad9429c5c9f3/src/apps/viewer/app.ts#L419-L434)

```javascript
        // Get the model
        model = this.plugin.managers.structure.hierarchy.current.structures[0].model

       // Get the coords
        const data = params.coordinates.kind === 'coordinates-data'
                ? await plugin.builders.data.rawData({ data: params.coordinates.data, label: params.coordinatesLabel })
                : await plugin.builders.data.download({ url: params.coordinates.url, isBinary: params.coordinates.isBinary, label: params.coordinatesLabel });

            const provider = plugin.dataFormats.get(params.coordinates.format);
            coords = await provider!.parse(plugin, data);
        const trajectory = await plugin.build().toRoot()
            .apply(TrajectoryFromModelAndCoordinates, {
                modelRef: model.ref,
                coordinatesRef: coords.ref
            }, { dependsOn: [model.ref, coords.ref] })
            .commit();

        const preset = await plugin.builders.structure.hierarchy.applyPreset(trajectory, params.preset ?? 'default');
```

**Reply by hainm** [reply]
The question is how can I update the coords for a given model from 2D (or 1D) array of float?

**Reply by dsehnal** [reply]
One way I can think of is creating a custom transform that merges the Model with your custom cords which you insert into your state tree. 

Something like `WithCoordinates :: Structure -> Structure { coords: PD.Value<number[]>() }` which you then update with your new data. 

@arose has done something to this degree if I remember correctly, maybe he has other tips?

**Reply by hainm** [reply]
> One way I can think of is creating a custom transform that merges the Model with your custom cords which you insert into your state tree.
> 
> Something like `WithCoordinates :: Structure -> Structure { coords: PD.Value<number[]>() }` which you then update with your new data.
> 
> @arose has done something to this degree if I remember correctly, maybe he has other tips?



Thanks. @arose: please 👀

**Comment by hainm** [comment]
@dsehnal 

I've used [this function loadStructureFromData](https://github.com/hainm/molstarview/blob/5fcf4dd921a5d1a75bb100b0d5ebcbf70a03ea35/js/src/widget.ts#L65-L70)  to load the pdb from a blob of string. Then I download the molstar state (.molj). However, the state does not seem to include the content of the pdb file I've loaded.

Do I misunderstood about the meaning of `.molj` file? or it's still experimental or something else? thanks.

**Reply by hainm** [reply]
# State 
My aim is the follow `nglview`: play something in Jupyter notebook -> save to html file and share the view to someone:

https://github.com/nglviewer/nglview/pull/687

**Reply by dsehnal** [reply]
It works for me:

```ts
const data = await fetch('https://cs.litemol.org/5ire/full');
const str = await data.text();
viewer.loadStructureFromData(str, 'mmcif');
```

![image](https://user-images.githubusercontent.com/21145776/173011887-043d948a-31e0-4ea3-95cf-47f80c4adc30.png)

**Reply by hainm** [reply]
Thanks David and sorry for the noise. I can see the structure in the `.molj` now.

**Comment by hainm** [comment]
# Fullscreen

@dsehnal Does `molstar` have fullscreen mode? If not, do you have a plan to support it? 
(By the way, I am using the [adapted code here](https://github.com/hainm/fullscreen/blob/master/js/src/fullscreen.ts) from NGL to trigger the fullscreen mode in Jupyter notebook.

e.g: Need to run `fs.fullscreen()`  in the cell 19 to trigger:
<img width="1146" alt="image" src="https://user-images.githubusercontent.com/4451957/173100124-a39eb6b2-0484-43af-a561-f40a9697f5e8.png">

### But it's always great the just to click a button in `molstar` view.

**Reply by dsehnal** [reply]
It has only "full window" mode, not fullscreen currently. But the full window + F11 should take care of that?

**Reply by hainm** [reply]
This is how the "full screen" looks like with/without the "full window" mode if using my code:

<img width="1344" alt="image" src="https://user-images.githubusercontent.com/4451957/173101944-710f8f5d-5901-456d-9df2-81460291a002.png">

<img width="1344" alt="image" src="https://user-images.githubusercontent.com/4451957/173101921-90018856-ed07-4d3d-a536-248da45525f1.png">

**Reply by hainm** [reply]
> But the full window + F11 should take care of that?

@dsehnal If the view is in Jupyter notebook, what happens if using F11? 
Here is from my testing with full window + F11 (there is an ugly text overlapping but I will take care that)

<img width="1338" alt="image" src="https://user-images.githubusercontent.com/4451957/173102488-b6268cc8-6e65-4d60-86e0-e0db67194946.png">

**Reply by dsehnal** [reply]
Yes, this requires a few CSS hacks to work correctly in Jupyter. If the mode is "expanded", setting 

```css
#jp-bottom-pannel, .p-TabBar, .lm-TabBar { visibility: hidden }
.jp-Notebook { overflow: hidden }
```

and also increasing the z-index of the original container.

But the NGL approach is nice too, might try adding that to Mol* later too I guess.

**Reply by hainm** [reply]
> But the NGL approach is nice too, might try adding that to Mol* later too I guess.

That would be nice. Thanks. I know `nglview` users like the fullscreen mode much.  (e.g: https://github.com/RosettaCommons/pyrosetta_viewer3d/pull/4#issuecomment-1149111290)

**Comment by hainm** [comment]
# Save and load state

Hi @dsehnal,

Saying I am having `this.plugin`, what' the idiom to save the state (e.g: this.model.molstarState = ...) and then reload it to recreate the existing view? Basically I don't want to download the `molj` file but want to use its content directly and to restore the view from it. 

PS: I've read https://github.com/molstar/molstar/blob/0cb2e5857abf112d675087c947afbf672107a29f/src/mol-plugin-state/manager/snapshots.ts and https://github.com/molstar/molstar/blob/d187757bbcbd355526dd29278ded0cebbd03b7fc/src/examples/proteopedia-wrapper/index.ts

**Reply by dsehnal** [reply]
``await this.plugin.state.setSnapshot(...)`` should do the trick.

**Reply by hainm** [reply]
> `await this.plugin.state.setSnapshot(...)` should do the trick.

thanks David. Works beautifully. I use that `getSnapshot/setSnapshot` so I could display the `view` multiple times in the notebook

<img width="589" alt="image" src="https://user-images.githubusercontent.com/4451957/173191509-60de7633-167c-4dc3-b84b-0bac49ab26fc.png">

**Comment by hainm** [comment]
# Representations

Dear @dsehnal 
I have at least two questions regarding this issue (and thank you so much for you helps). 

1. How can I add a structure without any representation (e.g: user can add it later).
2. If I use below code to remove the representation and then add the new one, I can not zoom in the structure anymore
```
    removeRepresentation(modelIndex){
        var st = this.plugin.managers.structure.hierarchy.current.structures[modelIndex]
        this.plugin.managers.structure.component.removeRepresentations(st.components)
    },

function addRepresentation(plugin, params, modelIndex){
    var st = plugin.managers.structure.hierarchy.current.structures[modelIndex]
    console.log("Calling from addRepresentation", st, params, modelIndex)
    var components = st.components
    plugin.dataTransaction(async () => {
        for (const component of components) {
            await plugin.builders.structure.representation.addRepresentation(
                component.cell, params)
        }
     })
}
```

Before removing:
<img width="998" alt="image" src="https://user-images.githubusercontent.com/4451957/173698770-9c707ca7-73ee-46ca-8f31-d4300055a287.png">

After removing and add new. (I don't change the view at all)
<img width="1007" alt="image" src="https://user-images.githubusercontent.com/4451957/173698897-4a1c0e4b-4eb8-48a4-a6c8-116e190a9bcc.png">

**Reply by hainm** [reply]
`PluginCommands.Camera.Reset(this.plugin, {})`  helps but the camera view is reset (as the command does). However, this is still unexpected since the original view is lost.

**Reply by dsehnal** [reply]
Yo,

at least two answers you shall receive then :)

To create only structure you need to add the parts "by hand", e.g.:

```ts
const data = await this.plugin.builders.data.rawData({ data: input.moleculeSdf }, { state: { isGhost: true } });
const trajectory = await this.plugin.builders.structure.parseTrajectory(data, 'mol');
const model = await this.plugin.builders.structure.createModel(trajectory);
const structure = await this.plugin.builders.structure.createStructure(model);

const all = await this.plugin.builders.structure.tryCreateComponentStatic(structure, 'all');
if (all) await this.plugin.builders.structure.representation.addRepresentation(all, { type: 'ball-and-stick', color: 'element-symbol', colorParams: { carbonColor: { name: 'element-symbol', params: {} } } });
```

Can also use ``applyPreset`` on the structure to make it behave like the "default" behavior or you can write your own presets:

```ts
await this.plugin.builders.structure.representation.applyPreset(structureProperties, 'auto');
```

--------------

With camera you have 2 options:

- Take and restore snapshot:

```ts
const snapshot = plugin.canvas3d.camera.getSnapshot();
...
plugin.canvas3d.requestCameraReset({ snapshot, durationMs: 0 });
```

- Set camera to manual mode where you always need to call ``Reset/setSnapshot`` by hand:

```ts
plugin.canvas3d.setProps({ camera: { manualReset: true } });
```

**Reply by hainm** [reply]
Thanks @dsehnal for your replies.

(1) I am able to add an structure without representation (and then I can add/remove representation)
(2) I have not tried the camera trick yet but using (1) approach, I can zoom in/out the structure without issue.

I think I start understanding about `molstar` now. Thanks again.

**Reply by dsehnal** [reply]
In general, the state is built though ``plugin.build()`` and applying a series of transforms and updates. This is what creates the "JSON" description of the state which is the "reconciled" similar to how react works when updating the DOM.
